### ISSUE TITLE
Pretrain for Vietnamese

### ISSUE THREAD

**Issue by v-nhandt21** [tranctan/Speaker-Embeddings#1]
Hi Tan, I am working on my thesis on TTS, I found this is amazing, I am going to use speaker embeddings as one of feature for my experiment.
Do you have avaiable Checkpoints for Vietnamese, can you share it.
Thank you Tan <3

**Comment by tranctan** [comment]
Hi thank you for your interest! This is just my practice repository and is not finished yet (I've been being lame). For fully functioning repository, please refer to the original one [here](https://github.com/resemble-ai/Resemblyzer).

In my experience, although the pretrained model in the above repository is not trained on Vietnamese, it gives reasonably good results on Vietnamese, you can start from there. Some interesting information on how to train on your own language is discussed [here](https://github.com/resemble-ai/Resemblyzer/issues/41).

I personally do not own any Vietnamese multi-speaker dataset, so I don't have any checkpoint on VNese. But there is a Zalo AI Competition 2020 that has speaker verification task (and of course with dataset), you can check if you can use it for research purpose or not.

**Comment by v-nhandt21** [comment]
Thank you Tan, I have tried to use the pretrain in English, but I do not know for sure whether it performs well or not, have you evaluate it with any metric?
Now I am training by using one-hot for each speaker as the baseline.

**Comment by tranctan** [comment]
In order to evaluate how well the model perform, I used 2 ways:
- Unsupervised: Visualizing clusters of speakers with corresponding labels to see if the same speaker samples are actually closed to each other. Follow [this example](https://github.com/resemble-ai/Resemblyzer/blob/master/demo03_projection.py)
- Supervised: Input: a pair of utterances - Output: Prediction of whether these utterances belong to a single speaker. The metric for this case is Accuracy. 

Notice that the model performs quite poorly for too-short utterances (utterances with only < 1s will be falsely classified and stay away from the samples even from the same speaker).

Hope this helps.

**Comment by v-nhandt21** [comment]
> In order to evaluate how well the model perform, I used 2 ways:
> 
> * Unsupervised: Visualizing clusters of speakers with corresponding labels to see if the same speaker samples are actually closed to each other. Follow [this example](https://github.com/resemble-ai/Resemblyzer/blob/master/demo03_projection.py)
> * Supervised: Input: a pair of utterances - Output: Prediction of whether these utterances belong to a single speaker. The metric for this case is Accuracy.
> 
> Notice that the model performs quite poorly for too-short utterances (utterances with only < 1s will be falsely classified and stay away from the samples even from the same speaker).
> 
> Hope this helps.

The second way seem as of ZaloAi challenge. Did you use this method for that task and what the acccuracy you got🥰

**Comment by tranctan** [comment]
Nice catch! It's worth learning the defined metrics from qualified competitions and yes, it's from Zalo. 

I did some tips and tricks with preprocessing+postprocessing, continue training the pretrained model with the competition dataset and got 0.87 (the pretrained alone was 0.75 as I remember). The competition data contains many background noises, If I have more time to handle that more precisely then I think the result will significantly be boosted up.

**Comment by v-nhandt21** [comment]
> Nice catch! It's worth learning the defined metrics from qualified competitions and yes, it's from Zalo.
> 
> I did some tips and tricks with preprocessing+postprocessing, continue training the pretrained model with the competition dataset and got 0.87 (the pretrained alone was 0.75 as I remember). The competition data contains many background noises, If I have more time to handle that more precisely then I think the result will significantly be boosted up.

Quow, amazing experiment, it seems verification problem is language dependency. I found that for Vietnamese there are Vivos and the other one is of Vinbigdata. As your instruction, I think i would try to retrain this on Vietnamese dataset.

**Comment by tranctan** [comment]
Glad that I can help. 

Some insights that might help you: I tried training from scratch the model with VNese data only (including VIVOS and ZaloAI data), but it gives NaN loss after a few epochs. This may due to the size of data is too small, that's why I decided to use the pretrained model and re-train it on the VNese dataset.

I would suggest that you can start with running the pre-trained model on the test dataset to have a baseline score. Then any further training or tips and tricks should give a higher score, so that you can make sure you're pointing in the right direction. 

Good luck!

**Comment by v-nhandt21** [comment]
> In order to evaluate how well the model perform, I used 2 ways:
> 
> * Unsupervised: Visualizing clusters of speakers with corresponding labels to see if the same speaker samples are actually closed to each other. Follow [this example](https://github.com/resemble-ai/Resemblyzer/blob/master/demo03_projection.py)
> * Supervised: Input: a pair of utterances - Output: Prediction of whether these utterances belong to a single speaker. The metric for this case is Accuracy.
> 
> Notice that the model performs quite poorly for too-short utterances (utterances with only < 1s will be falsely classified and stay away from the samples even from the same speaker).
> 
> Hope this helps.

Tan ơi, Which distant should I use to measure two speaker embeded vector is the same or not, is cosine similarity ok?

**Comment by tranctan** [comment]
Yes it should be cosine similarity. Assuming that you have `utterance_embed_1` and `utterance_embed_2`, cosine similarity is simply: `np.dot(utterance_embed_1, utterance_embed_2)`
